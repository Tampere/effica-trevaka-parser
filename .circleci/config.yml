# SPDX-FileCopyrightText: 2021 City of Tampere
#
# SPDX-License-Identifier: LGPL-2.1-or-later

version: 2.1

orbs:
  node: circleci/node@5.2.0

executors:
  reuse_tool:
    docker:
      - image: fsfe/reuse:0.13.0 # NOTE: Update bin/add-license-headers.sh to match
    resource_class: small
  machine:
    machine:
      image: ubuntu-2204:2024.01.1
    resource_class: medium
    environment:
      BUILDKIT_PROGRESS: plain

commands:
  checkout_with_submodules:
    steps:
      - checkout
      - run: git submodule sync
      - run: git submodule update --init --recursive

jobs:
  check_licenses:
    executor: reuse_tool
    steps:
      - checkout
      - run:
          name: reuse lint
          command: |
            reuse lint || {
              echo '';
              echo '=========================================';
              echo 'Please run: ./bin/add-license-headers.sh';
              echo '=========================================';
              exit 1
            }
  test:
    executor: machine
    steps:
      - checkout_with_submodules
      - run:
          name: Start up compose stack
          command: docker-compose -f docker-compose-it-db.yml up -d --wait --build
      - run:
          name: Waiting for compose stack to be up...
          command: docker-compose -f docker-compose-it-db.yml wait flyway
      - node/install-packages:
          include-branch-in-cache-key: false
      - run: npm run integration-test
      - run:
          name: Collect docker-compose logs
          command: docker-compose -f docker-compose-it-db.yml logs --tail=all > /tmp/docker-compose-logs.txt
          when: always
      - store_artifacts:
          path: /tmp/docker-compose-logs.txt
          destination: docker-compose-logs.txt

workflows:
  version: 2
  test:
    jobs:
      - check_licenses
      - test
