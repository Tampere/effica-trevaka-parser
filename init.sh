#!/bin/bash

# SPDX-FileCopyrightText: 2023-2024 Tampere region
#
# SPDX-License-Identifier: LGPL-2.1-or-later

set -e
set -u

function create_database() {
  psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" <<-EOSQL
    CREATE DATABASE $1;
    \connect $1
    CREATE SCHEMA migration;
EOSQL
}

create_database vesilahti
