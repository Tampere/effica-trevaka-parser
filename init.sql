-- SPDX-FileCopyrightText: 2021 City of Tampere
--
-- SPDX-License-Identifier: LGPL-2.1-or-later

-- Add extensions here as needed (for UUIDs at least)
CREATE SCHEMA IF NOT EXISTS ext;
CREATE SCHEMA IF NOT EXISTS migration;

CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA ext;
CREATE EXTENSION IF NOT EXISTS btree_gist WITH SCHEMA ext;