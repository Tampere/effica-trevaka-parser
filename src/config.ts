export const config = {
    cityVariant: process.env.CITY_VARIANT ?? "tampere",
    port: process.env.PORT ?? 3000,
    isTimed: process.env.ISTIMED?.toUpperCase() === "TRUE",
    migrationSchema: process.env.MIGRATION_SCHEMA ?? "migration",
    extensionSchema: process.env.EXTENSION_SCHEMA ?? "ext",
    migrationDb: {
        host: process.env.PGHOST ?? "localhost",
        port: 5432,
        user: process.env.PGUSER ?? "postgres",
        password: process.env.PGPASSWORD ?? "postgres",
        database: process.env.PGDATABASE ?? "migration",
    },
    xmlParserOptions: {
        parseNodeValue: true,
        arrayMode: true,
        trimValues: true,
        parseTrueNumberOnly: true
    },
    csvParserOptions: {
        trim: true,
        delimiter: "|"
    }
}