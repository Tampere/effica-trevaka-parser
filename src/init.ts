// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import migrationDb from "./db/db";
import { baseQueryParameters, runQueryFile } from "./util/queryTools";

export const initDb = async () => {
    return await migrationDb.tx(async (t) => {
        await runQueryFile("init.sql", t, baseQueryParameters);
        await runQueryFile("functions.sql", t, baseQueryParameters);
    });
};
