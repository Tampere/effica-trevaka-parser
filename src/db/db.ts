import pgPromise from "pg-promise"
import { config } from "../config"

export const pgp = pgPromise({})

// date parsing seems to be scuffed, does local tz transformation but declares it UTC -> corrupts dates
//pgp.pg.types.setTypeParser(1082, dateParser)

const migrationDb = pgp({ ...config.migrationDb })

export default migrationDb