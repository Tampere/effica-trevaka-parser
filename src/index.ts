// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import app from "./app"
import { config } from "./config"

app.listen(config.port, () => {
    console.log(`Parser app listening at http://localhost:${config.port}`)
})

