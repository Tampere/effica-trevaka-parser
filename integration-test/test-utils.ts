// SPDX-FileCopyrightText: 2023-2024 Tampere region
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import { pgp } from "../src/db/db";
import { importFilesFromDir } from "../src/io/io";
import { IDatabase, ITask } from "pg-promise";
import { IClient } from "pg-promise/typescript/pg-subset";
import { Config } from "../src/config";
import { transform } from "../src/transform";
import { transfer } from "../src/transfer";
import { baseQueryParameters } from "../src/util/queryTools";
import { CITY_SPECIFIC_MAPPINGS } from "../src/mapping/citySpecific";

export const AREA_COLUMN_SET = new pgp.helpers.ColumnSet(["id", "name"], {
    table: "care_area",
});

export const UNIT_COLUMN_SET = new pgp.helpers.ColumnSet(
    ["id", "name", "care_area_id", "provider_type"],
    {
        table: "daycare",
    },
);

export const imports = (
    db: IDatabase<{}, IClient>,
    ...options: { path: string; importTarget: string }[]
): Promise<any> =>
    db.tx((tx) =>
        Promise.all(
            options.map(({ path, importTarget }) =>
                importFilesFromDir(tx, {
                    path,
                    returnAll: false,
                    importTarget,
                }),
            ),
        ),
    );

export const transforms = (
    db: IDatabase<{}, IClient>,
    config: Config,
    ...args: string[]
) => db.tx((tx) => Promise.all(args.map((arg) => transform(tx, arg, config))));

export const transfers = (db: IDatabase<{}, IClient>, ...args: string[]) =>
    db.tx((tx) => Promise.all(args.map((arg) => transfer(tx, arg))));

export const findPlacementMappings = async (
    tx: ITask<{}>,
    cityVariant: string,
) => {
    const municipal = await tx.many(
        `
            SELECT mapping.key AS effica,
                   jsonb_build_object(
                           'id', service_need_option.id,
                           'name_fi', service_need_option.name_fi,
                           'valid_placement_type', service_need_option.valid_placement_type,
                           'default_option', service_need_option.default_option
                   )           AS evaka
            FROM jsonb_each($(placementMapping)) mapping
                     LEFT JOIN service_need_option ON service_need_option.id = (mapping.value ->> 'serviceNeedOptionId')::uuid
        `,
        CITY_SPECIFIC_MAPPINGS[cityVariant],
    );
    const voucher = await tx.many(
        `
            SELECT mapping.key AS effica,
                   jsonb_build_object(
                           'id', coalesce(voucher.id, municipal.id),
                           'name_fi', coalesce(voucher.name_fi, municipal.name_fi),
                           'valid_placement_type',
                           coalesce(voucher.valid_placement_type, municipal.valid_placement_type),
                           'default_option', coalesce(voucher.default_option, municipal.default_option)
                   )           AS evaka
            FROM jsonb_each($(placementMapping)) mapping
                     LEFT JOIN service_need_option voucher ON voucher.id =
                                                              (mapping.value ->> 'privateServiceVoucherServiceNeedOptionId')::uuid
                     LEFT JOIN service_need_option municipal ON municipal.id = (mapping.value ->> 'serviceNeedOptionId')::uuid
        `,
        CITY_SPECIFIC_MAPPINGS[cityVariant],
    );
    return { municipal, voucher };
};

export const findTransformPlacements = (tx: ITask<{}>) =>
    tx.manyOrNone(
        `
            SELECT effica_ssn,
                   placement_type,
                   placement_unit_id,
                   placement_start_date::text,
                   placement_end_date::text,
                   placement_place_guarantee,
                   service_need_option_id,
                   service_need_start_date::text,
                   service_need_end_date::text,
                   service_need_shift_care
            FROM $(migrationSchema:name).evaka_placement
            ORDER BY effica_ssn, placement_start_date, placement_type, service_need_start_date;
        `,
        { ...baseQueryParameters },
    );

export const findTransferPlacements = (tx: ITask<{}>) =>
    tx.manyOrNone(
        `
            SELECT person.social_security_number AS evaka_ssn,
                   placement.type                AS placement_type,
                   placement.unit_id             AS placement_unit,
                   placement.start_date::text    AS placement_start_date,
                   placement.end_date::text      AS placement_end_date,
                   placement.place_guarantee     AS placement_place_guarantee,
                   service_need.option_id        AS service_need_option_id,
                   service_need.start_date::text AS service_need_start_date,
                   service_need.end_date::text   AS service_need_end_date,
                   service_need.shift_care       AS service_need_shift_care
            FROM placement
                     JOIN person ON person.id = placement.child_id
                     LEFT JOIN service_need ON service_need.placement_id = placement.id
            ORDER BY evaka_ssn, placement_start_date, placement_type, service_need_start_date;
        `,
        { ...baseQueryParameters },
    );

export const findTransferIncomes = (tx: ITask<{}>) =>
    tx.manyOrNone(
        `
            SELECT person.social_security_number AS evaka_ssn,
                   data,
                   valid_from::text,
                   valid_to::text,
                   notes,
                   effect,
                   is_entrepreneur,
                   works_at_echa
            FROM income
                     JOIN person ON person.id = income.person_id
            ORDER BY evaka_ssn, valid_from
        `,
    );

export const findTransferFridge = (tx: ITask<{}>) =>
    tx.manyOrNone(
        `
            SELECT child.social_security_number         AS child_social_security_number,
                   head_of_child.social_security_number AS head_of_child_social_security_number,
                   fridge_child.start_date::text,
                   fridge_child.end_date::text
            FROM fridge_child
                     JOIN person child ON child.id = fridge_child.child_id
                     JOIN person head_of_child ON head_of_child.id = fridge_child.head_of_child
            ORDER BY child_social_security_number, head_of_child_social_security_number, start_date
        `,
    );
