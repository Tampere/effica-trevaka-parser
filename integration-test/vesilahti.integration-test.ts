// SPDX-FileCopyrightText: 2023-2024 Tampere region
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import { initDb } from "../src/init";
import { config as configFromEnv } from "../src/config";
import { pgp } from "../src/db/db";
import { cleanupDb } from "../src/util/testTools";
import { findPlacementMappings } from "./test-utils";

const config = {
    ...configFromEnv,
    cityVariant: "vesilahti",
    migrationDb: {
        ...configFromEnv.migrationDb,
        database: "evaka_vesilahti_local",
    },
};
const db = pgp(config.migrationDb);

beforeAll(() => db.tx(initDb));
beforeEach(() => db.tx(cleanupDb));
afterAll(db.$pool.end);

describe("vesilahti", () => {
    it("all placement mappings exists", async () => {
        const mappings = await db.tx((tx) =>
            findPlacementMappings(tx, config.cityVariant),
        );
        expect(mappings).toMatchSnapshot();
    });
});
