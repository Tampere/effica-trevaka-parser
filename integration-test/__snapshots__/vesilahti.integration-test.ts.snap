// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vesilahti all placement mappings exists 1`] = `
{
  "municipal": [
    {
      "effica": "Vuorohoito 121-140h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Vuorohoito 141-155h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Vuorohoito 156-170h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Vuorohoito yli 170 h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 0-90h",
      "evaka": {
        "default_option": false,
        "id": "50358394-b961-11eb-b51f-67ac436e5637",
        "name_fi": "Varhaiskasvatus 85 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 91-120h",
      "evaka": {
        "default_option": false,
        "id": "86ef70a0-bf85-11eb-91e6-1fb57a101165",
        "name_fi": "Varhaiskasvatus 86–120 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 121-140h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 141-155h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 156-170h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus yli 170h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
  ],
  "voucher": [
    {
      "effica": "Vuorohoito 121-140h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Vuorohoito 141-155h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Vuorohoito 156-170h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Vuorohoito yli 170 h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 0-90h",
      "evaka": {
        "default_option": false,
        "id": "50358394-b961-11eb-b51f-67ac436e5637",
        "name_fi": "Varhaiskasvatus 85 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 91-120h",
      "evaka": {
        "default_option": false,
        "id": "86ef70a0-bf85-11eb-91e6-1fb57a101165",
        "name_fi": "Varhaiskasvatus 86–120 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 121-140h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 141-155h",
      "evaka": {
        "default_option": false,
        "id": "503590f0-b961-11eb-b520-53740af3f7ef",
        "name_fi": "Varhaiskasvatus 121-150 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus 156-170h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
    {
      "effica": "Varhaiskasvatus yli 170h",
      "evaka": {
        "default_option": false,
        "id": "503591ae-b961-11eb-b521-1fca99358eed",
        "name_fi": "Varhaiskasvatus yli 151 tuntia / kuukausi",
        "valid_placement_type": "DAYCARE",
      },
    },
  ],
}
`;
